### PR TITLE
Various fixes to tests infrastructure

### LINKED ISSUE #245: test_client_no_client_cert fails
I'm running the test suite.

```
===================================== test session starts =====================================
platform linux -- Python 3.7.2, pytest-4.3.1, py-1.8.0, pluggy-0.9.0
cachedir: .tox/py37-extras/.pytest_cache
rootdir: /home/test/custodia/custodia, inifile: tox.ini
collected 162 items                                                                           

tests/test_authenticators.py ..
tests/test_cli.py .....
tests/test_custodia.py ..........................................F.
tests/test_httpd.py .
tests/test_ipa.py ........................
tests/test_message_kem.py ....
tests/test_misc.py ...
tests/test_plugins.py ....
tests/test_secrets.py ..................................
tests/test_server.py ....
tests/test_store.py ...
tests/test_store_sqlite.py ...........
tests/functional/test_basics.py ...
tests/functional/test_container.py .........
tests/functional/test_key.py ...........

========================================== FAILURES ===========================================
________________________ CustodiaHTTPSTests.test_client_no_client_cert ________________________

self = <test_custodia.CustodiaHTTPSTests testMethod=test_client_no_client_cert>

    def test_client_no_client_cert(self):
        client = CustodiaSimpleClient(self.socket_url + '/forwarder')
        client.headers['REMOTE_USER'] = 'test'
        client.set_ca_cert(self.ca_cert)
        # XXX workaround for requests bug with urllib3 v1.22
        with self.assertRaises(RequestsConnSSLErrors) as e:
            client.list_container('test')
        self.assert_ssl_error_msg("SSLV3_ALERT_HANDSHAKE_FAILURE",
>                                 e.exception)

tests/test_custodia.py:520: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/test_custodia.py:502: in assert_ssl_error_msg
    self.fail(str(exc))
E   AssertionError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))
```

It seems that urllib3 doesn't support TLS1.3 yet:
https://github.com/urllib3/urllib3/pull/1496

### LINKED ISSUE #247: Test suite fails against Pytest 5
I run the 'custodia' test suite with Pytest 5:
```
138 passed, 1 skipped, 1 warnings, 23 error
```
There are several identical errors:
```
==================================== ERRORS ====================================
_______________ ERROR at setup of TestBasics.test_default_answer _______________

self = <functional.test_basics.TestBasics object at 0x7fa1f58e4080>
simple_configuration = None
request = <SubRequest 'custodia_server' for <Function test_default_answer>>
dev_null = 4

    @pytest.fixture(scope="class")
    def custodia_server(self, simple_configuration, request, dev_null):
        # Don't write server messages to stdout unless we are in debug mode
        # pylint: disable=no-member
>       if pytest.config.getoption('debug') or \
                pytest.config.getoption('verbose'):
E               pytest.PytestDeprecationWarning: the `pytest.config` global is deprecated.  Please use `request.config` or `pytest_configure` (if you're a pytest plugin) instead.

tests/functional/base.py:101: PytestDeprecationWarning
```

### PR DESCRIPTION
**Pull request by simo5**
Fix issues cropping up with newer python tools versions
Fixes #245 
Fixes #247

### CODE REVIEW

**Comment by simo5** [comment]
@tiran PTAL

**Comment by simo5** [comment]
@stanislavlevin would you mind reviewing this PR given you proposed #248 and therefore are familiar with the issue ?

**Comment by stanislavlevin** [comment]
Please, let me check.

File: tests/test_custodia.py [hunk 1]
```diff
@@ -491,14 +491,16 @@ def create_client(cls, suffix, remote_user, clientcls=None):
...
 
-    def assert_ssl_error_msg(self, msg, exc):
-        # CERTIFICATE_VERIFY_FAILED, SSLV3_ALERT_HANDSHAKE_FAILURE
-        if msg in str(exc):
-            return
-        # 'certificate verify failed'
-        msg = msg.lower().replace('_', ' ')
-        if msg in str(exc):
-            return
+    def assert_ssl_error_msg(self, msgs, exc):
```
[the remark this answers is not in the document]
Comment (stanislavlevin) [reply tests/test_custodia.py]: I realized what is for.
If Python is linked with OpenSSL 1.1.1 urllib3 (at least 1.25.3) raises `ConnectionResetError: [Errno 104] Connection reset by peer`, while the proper one is `ssl.SSLError: [SSL] tlsv13 alert certificate required`.
This workaround also fixes #245.

**Comment by simo5** [comment]
@stanislavlevin thanks for the review, sorry for late reply, I was away

**Comment by simo5** [comment]
> ```diff
> -    if skip_servertest and item.get_marker("servertest") is not None:
> +    # pytest 4+
> +    if hasattr(item, 'get_closest_marker'):
> +        get_marker = item.get_closest_marker
> +    else:
> +        get_marker = item.get_marker
> +    if skip_servertest and get_marker("servertest") is not None:
> ```
> 
> You could use a simpler version of fix with support of pytest < 4 .

Do you have a suggestion ?

File: tests/functional/base.py [hunk 2]
```diff
@@ -98,8 +98,8 @@ def close_dev_null():
     def custodia_server(self, simple_configuration, request, dev_null):
         # Don't write server messages to stdout unless we are in debug mode
         # pylint: disable=no-member
-        if pytest.config.getoption('debug') or \
-                pytest.config.getoption('verbose'):
+        if request.config.getoption('debug') or \
```
[the remark this answers is not in the document]
Comment (simo5) [reply tests/functional/base.py]: ok

**Comment by stanislavlevin** [comment]
> > ```diff
> > -    if skip_servertest and item.get_marker("servertest") is not None:
> > +    # pytest 4+
> > +    if hasattr(item, 'get_closest_marker'):
> > +        get_marker = item.get_closest_marker
> > +    else:
> > +        get_marker = item.get_marker
> > +    if skip_servertest and get_marker("servertest") is not None:
> > ```
> > 
> > 
> > You could use a simpler version of fix with support of pytest < 4 .
> 
> Do you have a suggestion ?

That diff is the suggestion ;-)

**Comment by simo5** [comment]
Ok rebase with the requested changes, minus differences in how to handle the get_marker replacement

**Comment by stanislavlevin** [comment]
> > > ```diff
> > > -    if skip_servertest and item.get_marker("servertest") is not None:
> > > +    # pytest 4+
> > > +    if hasattr(item, 'get_closest_marker'):
> > > +        get_marker = item.get_closest_marker
> > > +    else:
> > > +        get_marker = item.get_marker
> > > +    if skip_servertest and get_marker("servertest") is not None:
> > > ```
> > > 
> > > 
> > > You could use a simpler version of fix with support of pytest < 4 .
> > 
> > 
> > Do you have a suggestion ?
> 
> Heh, nevermind, I just now realized this was the suggestion ... sometimes the github UI makes it confusing to distinguish actual code quoted from proposed code quoted ...
> 
> That said,
> I am not sure get_closest_marker gives me what I want, I had considered it and decided to check for the specific marker.
> But what I can do is retain the pytest < 4 code for backwards compat I guess.

Please, take a look at the official docs:
https://docs.pytest.org/en/latest/historical-notes.html#update-marker-code
http://doc.pytest.org/en/latest/example/markers.html#marking-whole-classes-or-modules

```python
for marker in item.iter_markers():
    if marker.name == "servertest":
        # args has --skip-servertests and test is marked as servertest
        pytest.skip("Skip integration test")
```

The iteration is extra operation here, because `iter_markers ` iterates over all markers of the item:
https://docs.pytest.org/en/latest/reference.html?highlight=get_closest_marker#_pytest.nodes.Node.iter_markers 
Instead, you could filter marks with `name=servertest`, but that is still returns `generator` of all the marks applied to the item (marks with the same name could be applied on test function, class or module levels). This can be done like:

```python
if skip_servertest:
    if next(item.iter_markers("servertest"), None) is not None:
        # args has --skip-servertests and test is marked as servertest
        pytest.skip("Skip integration test")
```

or you could just use the syntax sugar (`get_closest_marker`).

Internally, `get_closest_marker` calls `iter_markers(name=name)` and returns the first mark from generator.
https://docs.pytest.org/en/latest/_modules/_pytest/nodes.html#Node.get_closest_marker

Honestly speaking, the custodia test suite has only 1 mark and only one level for this mark, so, it doesn't matter how to handle this one :smile: But...

**Comment by stanislavlevin** [comment]
Another one point, what is about Python3.7 at Travis?
Like this:
[0001-py37.patch.txt](https://github.com/latchset/custodia/files/3515935/0001-py37.patch.txt)

**Comment by simo5** [comment]
Ok let me use get_closer_marker() and if later we have issues we'll fix em later...

As for py37 I think I tried to add it and something else blew up so I backed out :)

**Comment by simo5** [comment]
Pushed, let's see what happens

**Comment by simo5** [comment]
@stanislavlevin sounds like all checks pass, can you approve a review? (can't recall if permissions are set up that you can or not, let me know if you can't)

**Comment by simo5** [comment]
Sigh realized now I had opened this PR from a origin branch ... then kept pushing changes to my personal branch ... oooh well..

**Comment by stanislavlevin** [comment]
I've checked this PR locally against Python 3.7.3.
tox `py37` env is OK, unless `--skip-servertests`.
```
tox.py3 --sitepackages -e py37 -- --skip-servertests

platform linux -- Python 3.7.3, pytest-5.1.0, py-1.8.0, pluggy-0.12.0
cachedir: .tox/py37/.pytest_cache
rootdir: /usr/src/RPM/BUILD/test/custodia, inifile: tox.ini
collected 162 items                                                                           

tests/test_authenticators.py EE
tests/test_cli.py EEEEE
tests/test_custodia.py EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
tests/test_httpd.py E
tests/test_ipa.py EEEEEEEEEEEEEEEEEEEEEEEE
tests/test_message_kem.py EEEE
tests/test_misc.py EEE
tests/test_plugins.py EEEE
tests/test_secrets.py EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
tests/test_server.py EEEE
tests/test_store.py EEE
tests/test_store_sqlite.py EEEEEEEEEEE
tests/functional/test_basics.py EEE
tests/functional/test_container.py EEEEEEEEE
tests/functional/test_key.py EEEEEEEEEEE

=========================================== ERRORS ============================================
_______________________ ERROR at setup of TestAuthenticators.test_cred
...
>           elif item.get_closer_marker("servertest"):
E           AttributeError: 'TestCaseFunction' object has no attribute 'get_closer_marker'
```

The method should  be 'get_close**st**_marker', not 'get_close**r**_marker' ;-)

**Comment by simo5** [comment]
Oh my ... should be fixed now

**Comment by stanislavlevin** [comment]
Pytest | With Servertests | Python 2.7.16 | Python 3.7.3
--------- | ----------------------- | ------------------- | ------------------
3.10.1 |      Y | OK | OK
3.10.1 |     N | OK | OK
5.1.0 |       Y | OK | OK
5.1.0 |       N | OK | OK

**Comment by simo5** [comment]
Done,
thanks a lot for the review!

**Comment by stanislavlevin** [comment]
Thank you too!

**Pull request merged**